### DISCUSSION TITLE
Can I store encrypted files/secrets outside the main repository?

### DISCUSSION THREAD

**Discussion by Jackenmen** [twpayne/chezmoi#1512] [Q&A]
I use a public repository for my dotfiles and am wondering whether I can in some way store the encrypted files outside that repository while still benefitting from what chezmoi offers. Some things, I would want to store are my ssh key (`~/.ssh/id_rsa`) and tokens I'll be using in chezmoi templates (the templates will be part of the public repository). Even if encrypted, I would rather not put them in a repository accessible by anyone so I was hoping for a way to store it at a different not publically accessible place. Preferably another git repository (perhaps as a git submodule?) but if that's not achievable then a local directory/archive would be alright too as​ I can just manually get it onto the system before running chezmoi.

I don't really know whether I can achieve this with chezmoi already and am hoping to either find an already working solution or pointers on what feature I would have to request on the issue tracker to make this possible.

**Comment by twpayne** [comment]
For secrets like this, I'd strongly recommend [using a password manager](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md#keep-data-private) to store your secrets. That way the amount of information revealed in your public dotfiles repo is almost minimized. Here's [an example of a secret in my public dotfile repo that is stored outside the repository](https://github.com/twpayne/dotfiles/blob/master/dot_zshrc.tmpl#L99) - here I use a chezmoi template function to retrieve a secret from my personal password manager (LastPass), while still keeping my dotfiles repo public.

**Reply by Jackenmen** [reply]
I use Enpass as my password manager which does not have an official CLI + I currently am developing inside a Linux VM that runs on Windows *and* so far I have not felt the need to install the password manager inside it either which kind of stops me from using even the unofficial/3rd-party Enpass CLI.
And this doesn't really solve my issue with the storage of entire files like ssh keys.

This is why I was hoping to use a setup that does not involve my password manager at all if possible.

**Comment by twpayne** [comment]
Modern password managers can store entire files. Here's an [example from my personal dotfiles](https://github.com/twpayne/dotfiles/blob/master/private_dot_ssh/config.tmpl#L72) where I bring in a section of a config file from a password manager. Of course, this works for entire files too.

Modern encryption tools like [age](https://age-encryption.org/) and gpg combined with chezmoi's native support for them mean that you can push encrypted files to your public dotfile repo, [like this](https://github.com/zombiezen/dotfiles/blob/main/private_dot_ssh/encrypted_private_config.age). Note that chezmoi has builtin support for age.

I'm struggling to understand what exactly you want that isn't already provided by chezmoi. Would you be able to give a description of your ideal solution?

**Reply by Jackenmen** [reply]
> Modern password managers can store entire files. 

Enpass can too but as I said, there is no official CLI. The unofficial one is rather lacking, it does not seem to me like it can even get files from the vault. Either way, I am not interested in changing the password manager I already have and would in general prefer to not use a password manager for this as I sort of explained in the previous comment.

> I'm struggling to understand what exactly you want that isn't already provided by chezmoi.

I don't know if it's something that *isn't* provided by chezmoi, it could be that I missed something obvious. This is why I made this a discussion in Q&A section, not a trackable feature request yet.

> Would you be able to give a description of your ideal solution?

As I said in the original post, I don't have a specific solution in mind and am just, in general, looking for a way to store the encrypted files *outside* the main repository *but* have it work as if they were in the source state. Now based on your questions I feel I should add that I also do not want to use password manager, I did not think of including that when writing the post.

With that said, based on what I found in the chezmoi's documentation, I think I'm looking for something like the existing `.chezmoiexternal.<format>` but working with a git repository (that way I can have a second, private git repository with the encrypted files and secrets) or with a local file/archive rather than a one hosted at some URL (that way I don't need to have a running HTTP server *and* I don't have to have it be publically accessible). For the latter case, I can sort of workaround it already by temporarily running a web server locally but that does not seem ideal.
Perhaps a different solution than extending the `.chezmoiexternal.<format>` could also work but I haven't much more thought into what the ideal solution could be as I thought that perhaps you'll have some idea on what would be the best way to solve this.

**Reply by twpayne** [reply]
Thanks for the response :) https://github.com/twpayne/chezmoi/issues/1421 is possibly relevant.

**Reply by Jackenmen** [reply]
Yeah, that seems like something that would work for me, I'll subscribe to it.

**Reply by Jackenmen** [reply]
I've been thinking about this and wondered if it would be possible to use `Authorization` header/HTTP basic authentication (filled from template variables) when getting the external archive from a specified URL? I could then presumably [use GitHub's API](https://stackoverflow.com/a/23796159) to get a tarball of a private git repository.

**Reply by twpayne** [reply]
Yes, there's an example of this here: https://github.com/marcus-crane/dotfiles/blob/main/dot_ssh/config.tmpl#L12-L16

**Reply by Jackenmen** [reply]
> Yes, there's an example of this here: [marcus-crane/dotfiles@`main`/dot_ssh/config.tmpl#L12-L16](https://github.com/marcus-crane/dotfiles/blob/main/dot_ssh/config.tmpl#L12-L16)

Perhaps I'm missing something but I don't think this solves my issue.

What I want is to specify an archive URL in `.chezmoiexternal.<format>` file *and* specify `Authorization` header (or HTTP basic authentication) that would be used when chezmoi makes a HTTP request in order to download the archive.

To give a more concrete example, I want to have a `.chezmoiexternal.toml` file look somewhat like this (the structure I used for headers array/table doesn't matter, just wanted to quickly express it *in some way*):
```toml
[""]
    type = "archive"
    url = "https://github.com/jack1142/private-dotfiles/archive/main.tar.gz"
    headers = [
        {name = "Authorization", value = "{{ .name_of_the_template_variable_with_secret }}"}
    ]
    refreshPeriod = "168h"
```

---

I want to note here that I haven't actually tried adding `.chezmoiexternal.<format>` file at all yet so I am not actually entirely sure that it is even the right tool for the job here. I was hoping that I could leverage it to have the external archive be unpacked in the top-level directory of the source state *and* that I could use something like `.chezmoidata.<format>` (or files in `.chezmoitemplates` directory?) inside the archive to define the secrets to be used by other templates (templates that are part of the main dotfiles repository, not the archive) and also just put some whole files like SSH key(s).

I didn't really try looking into whether this actually works like that as for now I just got stuck on figuring out whether I can specify the `Authorization` header (or use HTTP basic authentication)

**Reply by marcus-crane** [reply]
> Yes, there's an example of this here: https://github.com/marcus-crane/dotfiles/blob/main/dot_ssh/config.tmpl#L12-L16

For historical purposes, the commit that was being pointed to lives here: https://github.com/marcus-crane/dotfiles/blob/b48dcd4/dot_ssh/config.tmpl#L11-L16

I've since changed how I deal with my work config :)

**Comment by twpayne** [comment]
By the way #741 was opened to add Enpass support to chezmoi, but there was no indication from Enpass that there would ever be an official CLI, so I closed the issue.

**Reply by Jackenmen** [reply]
Last time I searched, I found this response from Enpass team: https://discussion.enpass.io/index.php?/topic/4362-command-line-client/&tab=comments#comment-35116
I'm not sure whether they meant they're working on implementing it or if they're just "working" on whether they want to implement it and there's been no update in over a year so I don't have my hopes up. Due to my current setup, I don't *really* want to have Enpass on the target system at all since it's a VM and I already have the password manager on the host system but might still turn up to be useful in the future if they ever do implement it 😄
